utility/unit_of_work/uow_leader.go

```diff
@@ -0,0 +1,102 @@
+package unit_of_work
+
+import (
+	"github.com/pokt-network/pocket/logger"
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+var (
+	_ modules.UtilityUnitOfWork       = &leaderUtilityUnitOfWork{}
+	_ modules.LeaderUtilityUnitOfWork = &leaderUtilityUnitOfWork{}
+)
+
+type leaderUtilityUnitOfWork struct {
+	baseUtilityUnitOfWork
+}
+
+func NewForLeader(height int64, readContext modules.PersistenceReadContext, rwPersistenceContext modules.PersistenceRWContext) *leaderUtilityUnitOfWork {
+	return &leaderUtilityUnitOfWork{
+		baseUtilityUnitOfWork: baseUtilityUnitOfWork{
+			height:                 height,
+			persistenceReadContext: readContext,
```

Comment:
Why do we need a read context in addition to the write context?

Reply:
Separation of concerns. This way `readContext` could be a savepoint that implements the interface for example.
`writeContext` would be the real writecontext that is used on `Commit()` to persist the changes in UOW.

The fact that we have to compute the statehash from the UOW using the changes (still ephemeral) is an implementation detail, out of scope for this PR. It will implemented in #564